### ISSUE TITLE
False positive in Naming/MemoizedInstanceVariableName due to conditionally assigned instance variable

### ISSUE THREAD

**Issue by ChallahuAkbar** [rubocop-hq/rubocop#5836]
When a method implicitly returns an instance variable that is conditionally assigned, RuboCop detects it as memoization, so it triggers a false positive for the cop.

--------

## Expected behavior

No offense should be raised.

## Actual behavior

The implicitly returned instance variable is reported as a misnamed memoized variable.

## Steps to reproduce the problem

Simple example:
```ruby
class foo
  attr_accessor :problem_var

  def bar
    @problem_var ||= 'foobar'
  end
end
```

Where first observed: https://git.io/vp4Z7

## RuboCop version

```
$ rubocop -V
0.55.0 (using Parser 2.5.1.0, running on ruby 2.5.1 x86_64-linux)
```

**Comment by mikegee** [comment]
> When a method implicitly returns an instance variable that is conditionally assigned, RuboCop detects it as memoization

Indeed. How could Rubocop differentiate your example from actual memoization?

**Comment by ChallahuAkbar** [comment]
Well, as for the production example I gave, it's been "fixed" because we realized that we didn't need the conditional assignment line at all, but of course, that doesn't solve the larger issue. Thanks for showing how the simple example could be solved, though. As for how RuboCop could differentiate, in all honesty, I'm not totally sure; I don't have a lot of experience with this sort of thing. I just wanted to make sure that a possible bug didn't go unreported. I'm certainly interested in and willing to learn more in order to help though!

**Comment by mikegee** [comment]
I don't think this is a bug. When the code looks exactly the same as memoization from Rubocop's point of view, but the developer didn't intend it to be memoization, I think the only recourse is to disable the cop (either inline, or in the project's config file).

One way to make the code not look like memoization is to return some other value under the `@ivar ||= something` line.

**Comment by gaffneyc** [comment]
Adding to this issue, I just ran into this case where we needed to look up the given value or fallback to a known safe default.

```ruby
@thing   = Thing.find_by(name: params[:name])
@thing ||= Thing.find_by(name: "default")
```

In this case, if the call is overriding a variable is defined earlier in the method it should avoid triggering the cop.

**Comment by thegorgon** [comment]
Another example:

```
module AuditLog
  def self.setup
    @requests ||= {}
    @credentials ||= {}
  end
  ...
end
```

The presence of two `@variable ||= ...` patterns should be enough to exclude this cop, right?

**Comment by mikegee** [comment]
@gaffneyc I can't reproduce an offense for that code. Perhaps it was resolved since you commented.

@thegorgon That example warrants its own issue, in my opinion.

**Comment by ChallahuAkbar** [comment]
Yeah, this probably isn't something that can be "fixed" on rubocop's end. I'm closing the issue.

**Comment by thegorgon** [comment]
@mikegee - opened https://github.com/rubocop-hq/rubocop/issues/6271
